# Fortune magazine's 2017 Global 500

The pandas library provides solutions to all of these pain points and more. Pandas is not so much a replacement for NumPy as an extension of NumPy. The underlying code for pandas uses the NumPy library extensively, which means the concepts you've been learning will come in handy as you begin to learn more about pandas.

The primary data structure in pandas is called a dataframe. Dataframes are the pandas equivalent of a Numpy 2D ndarray, with a few key differences:

- Axis values can have string labels, not just numeric ones.
- Dataframes can contain columns with multiple data types: including integer, float, and string.

We'll work with a data set from Fortune magazine's 2017 Global 500 list, which ranks the top 500 corporations worldwide by revenue. The data set was originally compiled here; however, we modified the original data set to make it more accessible.

The data set is a CSV file called f500.csv. Here is a data dictionary for some of the columns in the CSV:

- company: Name of the company.
- rank: Global 500 rank for the company.
- revenues: Company's total revenue for the fiscal year, in millions of dollars (USD).
- revenue_change: Percentage change in revenue between the current and prior fiscal year.
- profits: Net income for the fiscal year, in millions of dollars (USD).
- ceo: Company's Chief Executive Officer.
- industry: Industry in which the company operates.
- sector: Sector in which the company operates.
- previous_rank: Global 500 rank for the company for the prior year.
- country: Country in which the company is headquartered.

We can use the DataFrame.dtypes attribute (similar to NumPy's ndarray.dtype attribute) to return information about the types of each column. Let's look at an example using a selection of data stored using the variable name f500_selection.

In [2]:
import pandas as pd
f500 = pd.read_csv('f500.csv',index_col=0)
f500.index.name = None
f500_type = type(f500)
f500_shape = f500.shape
f500_dtypes = f500.dtypes
print(f500.info())

<class 'pandas.core.frame.DataFrame'>
Index: 500 entries, Walmart to AutoNation
Data columns (total 16 columns):
rank                        500 non-null int64
revenues                    500 non-null int64
revenue_change              498 non-null float64
profits                     499 non-null float64
assets                      500 non-null int64
profit_change               436 non-null float64
ceo                         500 non-null object
industry                    500 non-null object
sector                      500 non-null object
previous_rank               500 non-null int64
country                     500 non-null object
hq_location                 500 non-null object
website                     500 non-null object
years_on_global_500_list    500 non-null int64
employees                   500 non-null int64
total_stockholder_equity    500 non-null int64
dtypes: float64(3), int64(7), object(6)
memory usage: 66.4+ KB
None


A summary of the techniques we've learned so far is below:

    Select by Label 	Explicit Syntax               Common Shorthand
    Single column       df.loc[:,"col1"]	          df["col1"]
    List of columns     df.loc[:,["col1", "col7"]]    df[["col1", "col7"]]
    Slice of columns	df.loc[:,"col1":"col4"]	

In [3]:
# select countries and assign

countries = f500['country']
countries

Walmart                               USA
State Grid                          China
Sinopec Group                       China
China National Petroleum            China
Toyota Motor                        Japan
                                   ...   
Teva Pharmaceutical Industries     Israel
New China Life Insurance            China
Wm. Morrison Supermarkets         Britain
TUI                               Germany
AutoNation                            USA
Name: country, Length: 500, dtype: object

### Select revenues and years_on_global_500_list

In [4]:
revenues_years = f500[['revenues', 'years_on_global_500_list']]
revenues_years

revenues  years_on_global_500_list
Walmart                           485873                        23
State Grid                        315199                        17
Sinopec Group                     267518                        19
China National Petroleum          262573                        17
Toyota Motor                      254694                        23
...                                  ...                       ...
Teva Pharmaceutical Industries     21903                         1
New China Life Insurance           21796                         2
Wm. Morrison Supermarkets          21741                        13
TUI                                21655                        23
AutoNation                         21609                        12

[500 rows x 2 columns]

### Select columns from ceo to sector

In [5]:
ceo_to_sector = f500.loc[:,'ceo':'sector']
ceo_to_sector

ceo  \
Walmart                         C. Douglas McMillon   
State Grid                                  Kou Wei   
Sinopec Group                             Wang Yupu   
China National Petroleum              Zhang Jianhua   
Toyota Motor                            Akio Toyoda   
...                                             ...   
Teva Pharmaceutical Industries    Yitzhak Peterburg   
New China Life Insurance                   Wan Feng   
Wm. Morrison Supermarkets            David T. Potts   
TUI                               Friedrich Joussen   
AutoNation                       Michael J. Jackson   

                                                       industry  \
Walmart                                   General Merchandisers   
State Grid                                            Utilities   
Sinopec Group                                Petroleum Refining   
China National Petroleum                     Petroleum Refining   
Toyota Motor                           Motor Vehicles and Parts   
...                                                         ...   
Teva Pharmaceutical Industries                  Pharmaceuticals   
New China Life Insurance        Insurance: Life, Health (stock)   
Wm. Morrison Supermarkets                  Food and Drug Stores   
TUI                                             Travel Services   
AutoNation                                  Specialty Retailers   

                                                sector  
Walmart                                      Retailing  
State Grid                                      Energy  
Sinopec Group                                   Energy  
China National Petroleum                        Energy  
Toyota Motor                    Motor Vehicles & Parts  
...                                                ...  
Teva Pharmaceutical Industries             Health Care  
New China Life Insurance                    Financials  
Wm. Morrison Supermarkets           Food & Drug Stores  
TUI                                  Business Services  
AutoNation                                   Retailing  

[500 rows x 3 columns]

### Observe all columns for Toyota

In [6]:
toyota = f500.loc['Toyota Motor']
toyota

rank                                                   5
revenues                                          254694
revenue_change                                       7.7
profits                                          16899.3
assets                                            437575
profit_change                                      -12.3
ceo                                          Akio Toyoda
industry                        Motor Vehicles and Parts
sector                            Motor Vehicles & Parts
previous_rank                                          8
country                                            Japan
hq_location                                Toyota, Japan
website                     http://www.toyota-global.com
years_on_global_500_list                              23
employees                                         364445
total_stockholder_equity                          157210
Name: Toyota Motor, dtype: object

### Create variable for drink companies

In [7]:
drink_companies = f500.loc[["Anheuser-Busch InBev", "Coca-Cola", "Heineken Holding"]]
drink_companies

rank  revenues  revenue_change  profits  assets  \
Anheuser-Busch InBev   206     45905             5.3   1241.0  258381   
Coca-Cola              235     41863            -5.5   6527.0   87270   
Heineken Holding       468     23044            -0.7    861.5   41469   

                      profit_change                        ceo   industry  \
Anheuser-Busch InBev          -85.0               Carlos Brito  Beverages   
Coca-Cola                     -11.2           James B. Quincey  Beverages   
Heineken Holding              -18.9  Jean-Francois van Boxmeer  Beverages   

                                         sector  previous_rank      country  \
Anheuser-Busch InBev  Food, Beverages & Tobacco            211      Belgium   
Coca-Cola             Food, Beverages & Tobacco            206          USA   
Heineken Holding      Food, Beverages & Tobacco            459  Netherlands   

                                 hq_location  \
Anheuser-Busch InBev         Leuven, Belgium   
Coca-Cola                        Atlanta, GA   
Heineken Holding      Amsterdam, Netherlands   

                                                website  \
Anheuser-Busch InBev            http://www.ab-inbev.com   
Coca-Cola               http://www.coca-colacompany.com   
Heineken Holding      http://www.theheinekencompany.com   

                      years_on_global_500_list  employees  \
Anheuser-Busch InBev                        12     206633   
Coca-Cola                                   23     100300   
Heineken Holding                            11      73525   

                      total_stockholder_equity  
Anheuser-Busch InBev                     71339  
Coca-Cola                                23062  
Heineken Holding                          6958

### Study middle_companies

In [8]:
middle_companies = f500.loc["Tata Motors":"Nationwide", "rank":"country"]
middle_companies

rank  revenues  revenue_change  profits  \
Tata Motors                       247     40329            -4.2   1111.6   
Aluminum Corp. of China           248     40278             6.0   -282.5   
Mitsui                            249     40275             1.6   2825.3   
Manulife Financial                250     40238            49.4   2209.7   
China Minsheng Banking            251     40234            -5.2   7201.6   
China Pacific Insurance (Group)   252     40193             2.2   1814.9   
American Airlines Group           253     40180            -2.0   2676.0   
Nationwide                        254     40074            -0.4    334.3   

                                 assets  profit_change                   ceo  \
Tata Motors                       42162          -34.0      Guenter Butschek   
Aluminum Corp. of China           75089            NaN              Yu Dehui   
Mitsui                           103231            NaN       Tatsuo Yasunaga   
Manulife Financial               537461           28.9     Donald A. Guloien   
China Minsheng Banking           848389           -1.8         Zheng Wanchun   
China Pacific Insurance (Group)  146873          -35.7          Huo Lianhong   
American Airlines Group           51274          -64.8     W. Douglas Parker   
Nationwide                       197790          -42.4  Stephen S. Rasmussen   

                                                                  industry  \
Tata Motors                                       Motor Vehicles and Parts   
Aluminum Corp. of China                                             Metals   
Mitsui                                                             Trading   
Manulife Financial                         Insurance: Life, Health (stock)   
China Minsheng Banking                       Banks: Commercial and Savings   
China Pacific Insurance (Group)            Insurance: Life, Health (stock)   
American Airlines Group                                           Airlines   
Nationwide                       Insurance: Property and Casualty (Mutual)   

                                                 sector  previous_rank country  
Tata Motors                      Motor Vehicles & Parts            226   India  
Aluminum Corp. of China                       Materials            262   China  
Mitsui                                      Wholesalers            245   Japan  
Manulife Financial                           Financials            394  Canada  
China Minsheng Banking                       Financials            221   China  
China Pacific Insurance (Group)              Financials            251   China  
American Airlines Group                  Transportation            236     USA  
Nationwide                                   Financials            241     USA

### Find the counts of each unique value in the country column in the dataframe.

In [9]:
countries = f500['country'].value_counts()
countries

USA             132
China           109
Japan            51
France           29
Germany          29
Britain          24
South Korea      15
Netherlands      14
Switzerland      14
Canada           11
Spain             9
Brazil            7
Italy             7
India             7
Australia         7
Taiwan            6
Russia            4
Ireland           4
Singapore         3
Sweden            3
Mexico            2
Israel            1
Indonesia         1
Finland           1
Belgium           1
Turkey            1
Venezuela         1
U.A.E             1
Norway            1
Saudi Arabia      1
Thailand          1
Denmark           1
Malaysia          1
Luxembourg        1
Name: country, dtype: int64

### Use the Series.value_counts() method to return the value counts for countries.

In [10]:
country_counts = countries.value_counts()
country_counts

1      13
7       4
4       2
29      2
14      2
3       2
51      1
109     1
24      1
15      1
11      1
9       1
6       1
132     1
2       1
Name: country, dtype: int64

### Look into which have been the biggest movers and create a new variable

In [11]:
big_movers = f500.loc[['Aviva','HP', 'JD.com','BHP Billiton'],['rank','previous_rank']]
big_movers

rank  previous_rank
Aviva           90            279
HP             194             48
JD.com         261            366
BHP Billiton   350            168

### Study the bottom companies and see if there is any interesting data points

In [12]:
bottom_companies = f500.loc['National Grid':'AutoNation', ['rank','sector','country']]
bottom_companies

rank              sector  country
National Grid                           491              Energy  Britain
Dollar General                          492           Retailing      USA
Telecom Italia                          493  Telecommunications    Italy
Xiamen ITG Holding Group                494         Wholesalers    China
Xinjiang Guanghui Industry Investment   495         Wholesalers    China
Teva Pharmaceutical Industries          496         Health Care   Israel
New China Life Insurance                497          Financials    China
Wm. Morrison Supermarkets               498  Food & Drug Stores  Britain
TUI                                     499   Business Services  Germany
AutoNation                              500           Retailing      USA

### Create new rank_change to see the difference

In [13]:
rank_change = f500['previous_rank'] - f500['rank']
rank_change

Walmart                             0
State Grid                          0
Sinopec Group                       1
China National Petroleum           -1
Toyota Motor                        3
                                 ... 
Teva Pharmaceutical Industries   -496
New China Life Insurance          -70
Wm. Morrison Supermarkets         -61
TUI                               -32
AutoNation                       -500
Length: 500, dtype: int64

### Find the maximum value for the rank_change series

In [14]:
rank_change_max = rank_change.max()
rank_change_max

226

### Find the minimum value for the rank_change series

In [15]:
rank_change_min = rank_change.min()
rank_change_min

-500

According to the data dictionary, this list should only rank companies on a scale of 1 to 500. Even if the company ranked 1st in the previous year moved to 500th this year, the rank change calculated would be -499. This indicates that there is incorrect data in either the rank column or previous_rank column.

### Look up descriptive statistics for the rank column in f500

In [16]:
rank_describe = f500['rank'].describe()
rank_describe

count    500.000000
mean     250.500000
std      144.481833
min        1.000000
25%      125.750000
50%      250.500000
75%      375.250000
max      500.000000
Name: rank, dtype: float64

If we use describe() on a column that contains non-numeric values, we get some different statistics.

In [17]:
country = f500["country"]
print(country.describe())

count     500
unique     34
top       USA
freq      132
Name: country, dtype: object


Earlier we used the Series.describe() method to explore the rank and previous_rank columns. The results had something odd - the minimum value for the previous_rank column is 0. However, this column should only have values between 1 and 500 (inclusive), so a value of 0 doesn't make sense. To investigate the possible cause of this issue, let's confirm the number of 0 values that appear in the previous_rank column.

### Confirm the number of 0 values in the previous_rank column

In [18]:
zero_previous_rank = f500['previous_rank'].value_counts().loc[0]
zero_previous_rank

33

We confirmed that 33 companies in the dataframe have a value of 0 in the previous_rank column. Given that multiple companies have a 0 rank, we might conclude that these companies didn't have a rank at all for the previous year. It would make more sense for us to replace these values with a null value instead.

### find the median (middle) value for the revenues and profits columns

In [19]:
#axis 0 give us the total for colum, if axis = 1 it would be the row
medians = f500[['revenues','profits']].median(axis=0) 
medians

revenues    40236.0
profits      1761.6
dtype: float64

### Find the maximum value for only the numeric columns from f500

In [20]:
max_f500 = f500.max(numeric_only=True)
max_f500

rank                            500.0
revenues                     485873.0
revenue_change                  442.3
profits                       45687.0
assets                      3473238.0
profit_change                  8909.5
previous_rank                   500.0
years_on_global_500_list         23.0
employees                   2300000.0
total_stockholder_equity     301893.0
dtype: float64

### return statistics for only object columns

In [21]:
objects = f500.describe(include=['O'])
objects

ceo                       industry      sector country  \
count               500                            500         500     500   
unique              500                             58          21      34   
top     Stefano Pessina  Banks: Commercial and Savings  Financials     USA   
freq                  1                             51         118     132   

           hq_location                   website  
count              500                       500  
unique             235                       500  
top     Beijing, China  http://www.sinopharm.com  
freq                56                         1

-----------------------------------------------------------------------
After reviewing the descriptive statistics for the numeric columns in f500, we can conclude that no values look unusual besides the 0 values in the previous_rank column. 

### Dow Chemical has a new CEO, update the value in the dataframe

In [23]:
f500['Dow Chemical', 'ceo'] = 'Jim Fitterling'

### Boolean index companies in "Motor Vehicles and Parts"

In [26]:
#Create a boolean series, motor_bool

motor_bool = f500['industry'] =="Motor Vehicles and Parts"

#Use the motor_bool boolean series to index the country column.

motor_countries = f500['country'][motor_bool]

### Use boolean indexing to update 0 values to NaN in the previous_rank column of the f500 dataframe

In [27]:
# check 0 values
prev_rank_before = f500['previous_rank'].value_counts(dropna=False).head()
print(prev_rank_before)

#value of np.nan where there previously was a value of 0.
f500.loc[f500['previous_rank']==0,'previous_rank'] = np.nan

# check Nan values
prev_rank_after = f500['previous_rank'].value_counts(dropna=False).head()
print(prev_rank_after)

0      33
159     1
147     1
148     1
149     1
Name: previous_rank, dtype: int64
NaN      33
471.0     1
234.0     1
125.0     1
166.0     1
Name: previous_rank, dtype: int64


### create a new rank_change column in our f500 dataframe

In [29]:
f500['rank_change'] = f500['previous_rank'] - f500['rank']

rank_change_describe = f500['rank_change'].describe()
rank_change_describe

count    467.000000
mean      -3.533191
std       44.293603
min     -199.000000
25%      -21.000000
50%       -2.000000
75%       10.000000
max      226.000000
Name: rank_change, dtype: float64

We can now verify that the minimum value of the rank_change column is now greater than -500

### Create a series, industry_usa, containing counts of the two most common values in the industry column for companies headquartered in the USA.

Expanded Version:
  
    usa = f500['country'] =='USA'
    industry_usa = f500['industry'][usa]
    industry_usa.value_counts().head(2)

In [33]:
industry_usa = f500['industry'][f500['country'] =='USA'].value_counts().head(2)
industry_usa

Banks: Commercial and Savings               8
Insurance: Property and Casualty (Stock)    7
Name: industry, dtype: int64

### Create a series, sector_china, containing counts of the three most common values in the sector column for companies headquartered in the China.

    china = f500['country'] == 'China'
    sector_china = f500['sector'][china]
    sector_china.value_counts().head(3)

In [34]:
sector_china = f500['sector'][f500['country']=='China'].value_counts().head(3)

### Select the first and seventh rows and the first five columns of the f500 dataframe.

In [35]:
first_seventh_row_slice = f500.iloc[[0,6],0:5]
first_seventh_row_slice

rank  revenues  revenue_change  profits  assets
Walmart               1    485873             0.8  13643.0  198825
Royal Dutch Shell     7    240033           -11.8   4575.0  411275

### select all rows from f500 that have a null value for the previous_rank column. Select only the company, rank, and previous_rank columns

    null = f500["previous_rank"].isnull()
    null_previous_rank = f500[null][["company","rank","previous_rank"]]

In [47]:
f500 = pd.read_csv("f500.csv")
f500.loc[f500['previous_rank']==0, "previous_rank"] = np.nan
# chained version

null_previous_rank = f500[f500["previous_rank"].isnull()][["company","rank", "previous_rank"]]
null_previous_rank

company  rank  previous_rank
48                   Legal & General Group    49            NaN
90                                  Uniper    91            NaN
123                      Dell Technologies   124            NaN
138                 Anbang Insurance Group   139            NaN
140                        Albertsons Cos.   141            NaN
180             Hewlett Packard Enterprise   181            NaN
267                           Hengli Group   268            NaN
271         Johnson Controls International   272            NaN
341                                  Chubb   342            NaN
375                 Charter Communications   376            NaN
376                              Medtronic   377            NaN
392                               Facebook   393            NaN
409                            Kraft Heinz   410            NaN
414                                  Nokia   415            NaN
431                          Standard Life   432            NaN
440            Brookfield Asset Management   441            NaN
441        Mercantil Servicios Financieros   442            NaN
446                            Samsung C&T   447            NaN
456                        Heraeus Holding   457            NaN
458                Yango Financial Holding   459            NaN
461                  Alibaba Group Holding   462            NaN
463                            Progressive   464            NaN
466                Country Garden Holdings   467            NaN
472                                 Altice   473            NaN
477                       Tencent Holdings   478            NaN
482                                  Aflac   483            NaN
484                  Suning Commerce Group   485            NaN
487                             Xiamen C&D   488            NaN
491                         Dollar General   492            NaN
493               Xiamen ITG Holding Group   494            NaN
494  Xinjiang Guanghui Industry Investment   495            NaN
495         Teva Pharmaceutical Industries   496            NaN
499                             AutoNation   500            NaN

### Create new column for not notnull() values from pervious_rank

In [50]:
# Use the Series.notnull() method to select all rows from f500 that have a non-null value for the previous_rank column.
previously_ranked = f500[f500['previous_rank'].notnull()]

#From the previously_ranked dataframe, subtract the rank column from the previous_rank column. 
rank_change = previously_ranked['previous_rank'] - previously_ranked['rank']

#Assign the values in the rank_change to a new column in the f500 dataframe, "rank_change".
f500['rank_change'] = rank_change

#inspect new column
f500.head(3)

company  rank  revenues  revenue_change  profits  assets  \
0        Walmart     1    485873             0.8  13643.0  198825   
1     State Grid     2    315199            -4.4   9571.3  489838   
2  Sinopec Group     3    267518            -9.1   1257.9  310726   

   profit_change                  ceo               industry     sector  \
0           -7.2  C. Douglas McMillon  General Merchandisers  Retailing   
1           -6.2              Kou Wei              Utilities     Energy   
2          -65.0            Wang Yupu     Petroleum Refining     Energy   

   previous_rank country      hq_location                 website  \
0            1.0     USA  Bentonville, AR  http://www.walmart.com   
1            2.0   China   Beijing, China  http://www.sgcc.com.cn   
2            4.0   China   Beijing, China  http://www.sinopec.com   

   years_on_global_500_list  employees  total_stockholder_equity  rank_change  
0                        23    2300000                     77798          0.0  
1                        17     926067                    209456          0.0  
2                        19     713288                    106523          1.0

### Select all companies with revenues over 100 billion and negative profits

In [53]:
# Select the companies with revenues greater than 100 billion
larger_revenue = f500['revenues'] > 100000

# Select the companies with profits less than 0.
negative_profits = f500['profits'] < 0

# Combine large_revenue and negative_profits
combined = larger_revenue & negative_profits

# Use combined to filter f500
big_rev_neg_profit = f500[combined]

big_rev_neg_profit

company  rank  revenues  revenue_change  profits   assets  \
32  Japan Post Holdings    33    122990             3.6   -267.4  2631385   
44              Chevron    45    107567           -18.0   -497.0   260078   

    profit_change               ceo                         industry  \
32         -107.5  Masatsugu Nagato  Insurance: Life, Health (stock)   
44         -110.8    John S. Watson               Petroleum Refining   

        sector  previous_rank country    hq_location                  website  \
32  Financials           37.0   Japan   Tokyo, Japan  http://www.japanpost.jp   
44      Energy           31.0     USA  San Ramon, CA   http://www.chevron.com   

    years_on_global_500_list  employees  total_stockholder_equity  rank_change  
32                        21     248384                     91532          4.0  
44                        23      55200                    145556        -14.0

### Select all rows for companies headquartered in either Brazil or Venezuela.

In [54]:
brazil_venezuela = f500[(f500['country'] == 'Brazil') | (f500['country'] =='Venezuela')]
brazil_venezuela

company  rank  revenues  revenue_change  profits  \
74                         Petrobras    75     81405           -16.3  -4838.0   
112            Itau Unibanco Holding   113     66876            21.4   6666.4   
150                  Banco do Brasil   151     58093           -13.4   2013.8   
153                   Banco Bradesco   154     57443            31.3   5127.9   
190                              JBS   191     48825            -0.1    107.7   
369                             Vale   370     29363            14.7   3982.0   
441  Mercantil Servicios Financieros   442     24403            50.3   2004.2   
486                Ultrapar Holdings   487     22167            -2.3    447.5   

     assets  profit_change                        ceo  \
74   246983            NaN       Pedro Pullen Parente   
112  415972          -13.7    Candido Botelho Bracher   
150  426416          -52.3   Paulo Rogerio Caffarelli   
153  366418           -5.7  Luiz Carlos Trabuco Cappi   
190   31605          -92.3    Wesley Mendonca Batista   
369   99014            NaN          Fabio Schvartsman   
441  148659          -10.5      Gustavo J. Vollmer A.   
486    7426           -0.8            Thilo Mannhardt   

                          industry                     sector  previous_rank  \
74              Petroleum Refining                     Energy           58.0   
112  Banks: Commercial and Savings                 Financials          159.0   
150  Banks: Commercial and Savings                 Financials          115.0   
153  Banks: Commercial and Savings                 Financials          209.0   
190                Food Production  Food, Beverages & Tobacco          185.0   
369   Mining, Crude-Oil Production                     Energy          417.0   
441  Banks: Commercial and Savings                 Financials            NaN   
486                         Energy                     Energy          474.0   

       country             hq_location                        website  \
74      Brazil  Rio de Janeiro, Brazil    http://www.petrobras.com.br   
112     Brazil       Sao Paulo, Brazil         http://www.itau.com.br   
150     Brazil        Brasilia, Brazil           http://www.bb.com.br   
153     Brazil          Osasco, Brazil     http://www.bradesco.com.br   
190     Brazil       Sao Paulo, Brazil  http://jbss.infoinvest.com.br   
369     Brazil  Rio de Janeiro, Brazil            http://www.vale.com   
441  Venezuela      Caracas, Venezuela             http://www.msf.com   
486     Brazil       Sao Paulo, Brazil        http://www.ultra.com.br   

     years_on_global_500_list  employees  total_stockholder_equity  \
74                         23      68829                     76779   
112                         4      94779                     37680   
150                        23     100622                     26551   
153                        21      94541                     32369   
190                         8     237061                      7307   
369                        11      73062                     39042   
441                         1       8370                      7550   
486                         8      15173                      2621   

     rank_change  
74         -17.0  
112         46.0  
150        -36.0  
153         55.0  
190         -6.0  
369         47.0  
441          NaN  
486        -13.0

### Select the first five companies in the Technology sector that are not headquartered in the USA 

In [56]:
tech_outside_usa = f500[(f500['sector'] =='Technology') &  ~(f500['country'] == 'USA')].head()
tech_outside_usa

company  rank  revenues  revenue_change  profits  \
14           Samsung Electronics    15    173957            -2.0  19316.5   
26    Hon Hai Precision Industry    27    135129            -4.3   4608.8   
70                       Hitachi    71     84558             1.2   2134.3   
82   Huawei Investment & Holding    83     78511            24.9   5579.4   
104                         Sony   105     70170             3.9    676.4   

     assets  profit_change                   ceo  \
14   217104           16.8          Oh-Hyun Kwon   
26    80436           -0.4             Terry Gou   
70    86742           48.8  Toshiaki Higashihara   
82    63837           -5.0          Ren Zhengfei   
104  158519          -45.1           Kazuo Hirai   

                                       industry      sector  previous_rank  \
14               Electronics, Electrical Equip.  Technology           13.0   
26               Electronics, Electrical Equip.  Technology           25.0   
70               Electronics, Electrical Equip.  Technology           79.0   
82   Network and Other Communications Equipment  Technology          129.0   
104              Electronics, Electrical Equip.  Technology          113.0   

         country              hq_location                 website  \
14   South Korea       Suwon, South Korea  http://www.samsung.com   
26        Taiwan  New Taipei City, Taiwan  http://www.foxconn.com   
70         Japan             Tokyo, Japan  http://www.hitachi.com   
82         China          Shenzhen, China   http://www.huawei.com   
104        Japan             Tokyo, Japan     http://www.sony.net   

     years_on_global_500_list  employees  total_stockholder_equity  \
14                         23     325000                    154376   
26                         13     726772                     33476   
70                         23     303887                     26632   
82                          8     180000                     20159   
104                        23     128400                     22415   

     rank_change  
14          -2.0  
26          -2.0  
70           8.0  
82          46.0  
104          8.0

### Find the company headquartered in Japan with the largest number of employees.
    selected_rows = f500[f500["country"] == "Japan"]
    sorted_rows = selected_rows.sort_values("employees", ascending=False)
    top_japanese_employer = sorted_rows.iloc[0]["company"]

In [69]:
top_jap_employer = f500[f500['country']=='Japan'].sort_values('employees',ascending=False).iloc[0]["company"]
top_jap_employer

'Toyota Motor'

### Produce the following dictionary of the top employer in each country

In [73]:
# Create an empty dictionary, top_employer_by_country to store the results of the exercise.
top_employer_by_country = {}

#Use the Series.unique() method to create an array of unique values from the country column
countries = f500['country'].unique()

#Use a for loop to iterate over the array unique countries.
for c in countries:
    # Select only the rows that have a country name equal to the current iteration
    selected_rows = f500[f500['country'] == c]
    # Use DataFrame.sort_values() to sort those rows by the employees column in descending order.
    sorted_rows = selected_rows.sort_values("employees", ascending=False)
    # Select the first row from the sorted dataframe.
    top_employer =sorted_rows.iloc[0]
    # Extract the company name from the index label company from the first row.
    name_top_employer = top_employer['company']
    # Assign the results to the top_employer_by_country dictionary
    top_employer_by_country[c] = name_top_employer

#Check dictionary
top_employer_by_country

{'USA': 'Walmart',
 'China': 'China National Petroleum',
 'Japan': 'Toyota Motor',
 'Germany': 'Volkswagen',
 'Netherlands': 'EXOR Group',
 'Britain': 'Compass Group',
 'South Korea': 'Samsung Electronics',
 'Switzerland': 'Nestle',
 'France': 'Sodexo',
 'Taiwan': 'Hon Hai Precision Industry',
 'Singapore': 'Flex',
 'Italy': 'Poste Italiane',
 'Russia': 'Gazprom',
 'Spain': 'Banco Santander',
 'Brazil': 'JBS',
 'Mexico': 'America Movil',
 'Luxembourg': 'ArcelorMittal',
 'India': 'State Bank of India',
 'Malaysia': 'Petronas',
 'Thailand': 'PTT',
 'Australia': 'Wesfarmers',
 'Belgium': 'Anheuser-Busch InBev',
 'Norway': 'Statoil',
 'Canada': 'George Weston',
 'Ireland': 'Accenture',
 'Indonesia': 'Pertamina',
 'Denmark': 'Maersk Group',
 'Saudi Arabia': 'SABIC',
 'Sweden': 'H & M Hennes & Mauritz',
 'Finland': 'Nokia',
 'Venezuela': 'Mercantil Servicios Financieros',
 'Turkey': 'Koc Holding',
 'U.A.E': 'Emirates Group',
 'Israel': 'Teva Pharmaceutical Industries'}

### Create a new column for ROA

In [76]:
f500['roa'] = f500['profits'] / f500['assets']

#check
f500.head(3)

company  rank  revenues  revenue_change  profits  assets  \
0        Walmart     1    485873             0.8  13643.0  198825   
1     State Grid     2    315199            -4.4   9571.3  489838   
2  Sinopec Group     3    267518            -9.1   1257.9  310726   

   profit_change                  ceo               industry     sector  \
0           -7.2  C. Douglas McMillon  General Merchandisers  Retailing   
1           -6.2              Kou Wei              Utilities     Energy   
2          -65.0            Wang Yupu     Petroleum Refining     Energy   

   previous_rank country      hq_location                 website  \
0            1.0     USA  Bentonville, AR  http://www.walmart.com   
1            2.0   China   Beijing, China  http://www.sgcc.com.cn   
2            4.0   China   Beijing, China  http://www.sinopec.com   

   years_on_global_500_list  employees  total_stockholder_equity  rank_change  \
0                        23    2300000                     77798          0.0   
1                        17     926067                    209456          0.0   
2                        19     713288                    106523          1.0   

        roa  
0  0.068618  
1  0.019540  
2  0.004048

### Produce the following dictionary top company in sector

In [78]:
top_roa_by_sector = {}

sector = f500['sector'].unique()

for x in sector:
    sector_name = f500['sector'] == x
    sector_companies = f500[sector_name]
    sort_values = sector_companies.sort_values("roa", ascending = False)
    top_company = sort_values.iloc[0]
    company_name = top_company['company']
    top_roa_by_sector[x]= company_name
    
top_roa_by_sector

{'Retailing': 'H & M Hennes & Mauritz',
 'Energy': 'National Grid',
 'Motor Vehicles & Parts': 'Subaru',
 'Financials': 'Berkshire Hathaway',
 'Technology': 'Accenture',
 'Wholesalers': 'McKesson',
 'Health Care': 'Gilead Sciences',
 'Telecommunications': 'KDDI',
 'Engineering & Construction': 'Pacific Construction Group',
 'Industrials': '3M',
 'Food & Drug Stores': 'Publix Super Markets',
 'Aerospace & Defense': 'Lockheed Martin',
 'Food, Beverages & Tobacco': 'Philip Morris International',
 'Household Products': 'Unilever',
 'Transportation': 'Delta Air Lines',
 'Materials': 'CRH',
 'Chemicals': 'LyondellBasell Industries',
 'Media': 'Disney',
 'Apparel': 'Nike',
 'Hotels, Restaurants & Leisure': 'McDonald’s',
 'Business Services': 'Adecco Group'}